### PR TITLE
[SNOW-1524894]: Add support for `reindex`

### PR DESCRIPTION
**Pull request by sfc-gh-rdurrani**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1524894

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

Add support for `reindex`.

### CODE REVIEW

File: docs/source/modin/supported/series_supported.rst [hunk 1]
```diff
@@ -186,7 +186,7 @@ Methods
 +-----------------------------+---------------------------------+----------------------------------+----------------------------------------------------+
 | ``factorize``               | N                               |                                  |                                                    |
 +-----------------------------+---------------------------------+----------------------------------+----------------------------------------------------+
-| ``ffill``                   | P                               |                                  | ``N`` if parameter ``downcast`` is set. ``limit``  |
+| ``ffill``                   | P                               | ``downcast``                     | ``N`` if parameter ``downcast`` is set. ``limit``  |
```
Comment (sfc-gh-dpetersohn) [inline docs/source/modin/supported/series_supported.rst:189]: Nit: Suggest to change in a new PR

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 2]
```diff
@@ -8812,32 +9049,80 @@ def _make_fill_expression_for_column_wise_fillna(
         col_pos = self._modin_frame.data_column_snowflake_quoted_identifiers.index(
             snowflake_quoted_identifier
         )
+
```
Comment (sfc-gh-dpetersohn) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:9052]:
```suggestion
```

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 3]
```diff
@@ -8812,32 +9049,80 @@ def _make_fill_expression_for_column_wise_fillna(
...
         # If we are looking at the first column and doing an ffill, or looking at the last column and doing a bfill,
         # there are no other columns for us to coalesce with, so returning coalesce will error since it will be a
         # coalesce with one column. Instead, we just return the column.
         if (col_pos == 0 and method_is_ffill) or (
             col_pos == len_ids - 1 and not method_is_ffill
         ):
             return col(snowflake_quoted_identifier)
+
```
Comment (sfc-gh-dpetersohn) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:9060]:
```suggestion
```

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 4]
```diff
@@ -8849,6 +9134,8 @@ def fillna(
         axis: Optional[Axis] = None,
         limit: Optional[int] = None,
         downcast: Optional[dict] = None,
+        _filter_column_snowflake_quoted_id: Optional[str] = None,
+        _columns_to_ignore: Optional[list[bool]] = None,
```
Comment (sfc-gh-dpetersohn) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:9138]: Is it possible to do this differently, i.e. not change the query compiler method's signature? I'm fine with doing this on internal methods that have leading underscores, but it'll be problematic if we do this for the public interface we want to inherit from `BaseQueryCompiler`.

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 5]
```diff
@@ -2176,6 +2176,242 @@ def any(
...
+        fill_value : scalar
+            Value to use for missing values in the resulted frame.
+        limit : int
+        tolerance : int
+        **kwargs : dict
+            Serves the compatibility purpose. Does not affect the result.
+
+        Returns
+        -------
+        BaseQueryCompiler
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:2205]: SnowflakeQueryCompiler?

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py [hunk 6]
```diff
@@ -2176,6 +2176,242 @@ def any(
...
+            QueryCompiler with aligned axis.
+        """
+        if axis == 0:
+            new_index_qc = pd.Series(labels)._query_compiler
```
Comment (sfc-gh-lspiegelberg) [inline src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:2209]: maybe split into two helpers? One for axis=0, one for axis=1? This method is very long :)

File: src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py (hunk 4 above)
Comment (sfc-gh-rdurrani) [reply src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py:9138]: Done! Moved the code to a new underscored method with the additional arguments, and call the new underscored method from fillna without the additional arguments!

**Comment by sfc-gh-helmeleegy** [comment]
The to_csv fix is already in. So if you merge with main, you should be able to get rid of some of the test failures.

File: tests/integ/modin/frame/test_reindex.py [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+import numpy as np
+import pandas as native_pd
+import pytest
+
+import snowflake.snowpark.modin.plugin  # noqa: F401
+from snowflake.snowpark.exceptions import SnowparkSQLException
+from tests.integ.modin.sql_counter import SqlCounter, sql_count_checker
+from tests.integ.modin.utils import eval_snowpark_pandas_result
+
+
```
Comment (sfc-gh-lmukhopadhyay) [inline tests/integ/modin/frame/test_reindex.py:15]: Should you add sql count checker here?

**Review by sfc-gh-lmukhopadhyay — COMMENTED** [review]
Not sure if I missed it, do we have negative MultiIndex tests for frame/series here?

File: tests/integ/modin/frame/test_reindex.py (hunk 7 above)
Comment (sfc-gh-rdurrani) [reply tests/integ/modin/frame/test_reindex.py:18]: Good catch, thank you!

**Review by sfc-gh-helmeleegy — APPROVED** [review]
LGTM. Thanks for the nice refactoring, the comments with all the clarifying examples and the thorough test coverage. Great job!

**Review by sfc-gh-lmukhopadhyay — APPROVED** [review]
LGTM!

**Pull request merged**